File: rust/src/leetcode/Q309_best_time_to_buy_and_sell_stock_with_cooldown.rs
/*
 * @lc app=leetcode id=309 lang=rust
 *
 * [309] Best Time to Buy and Sell Stock with Cooldown
 */

// @lc code=start
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut bought: Vec<i32> = vec![-prices[0]; prices.len()];
        let mut sold: Vec<i32> = vec![0; prices.len()];
        let mut chilled: Vec<i32> = vec![0; prices.len()];

        for i in 1..prices.len() {
            bought[i] = bought[i - 1].max(chilled[i - 1] - prices[i]);
            sold[i] = bought[i - 1] + prices[i];
            chilled[i] = chilled[i - 1].max(sold[i - 1]);
        }

        sold[sold.len() - 1].max(chilled[chilled.len() - 1])
    }
}
// @lc code=end

File: python/Q309_best_time_to_buy_and_sell_stock_with_cooldown.py
#
# @lc app=leetcode id=309 lang=python3
#
# [309] Best Time to Buy and Sell Stock with Cooldown
#


# @lc code=start
class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        buy: list[int] = [-prices[0] for _ in range(len(prices))]
        sell: list[int] = [0 for _ in range(len(prices))]
        cooldown: list[int] = [0 for _ in range(len(prices))]

        for i in range(1, len(prices)):
            buy[i] = max(buy[i - 1], cooldown[i - 1] - prices[i])
            sell[i] = buy[i - 1] + prices[i]
            cooldown[i] = max(sell[i - 1], cooldown[i - 1])

        return max(sell[-1], cooldown[-1])


# @lc code=end
